### DISCUSSION TITLE
Can Kitty distinguish between right and left keyboard modifiers?

### DISCUSSION THREAD

**Discussion by roscrl** [kovidgoyal/kitty#5331] [Q&A] [answered]
I have my hyper key set on capslock as `right control + right command + right shift + right option` using Karabiner Elements on MacOS.

This allows me to do things like `` Hyper + ` `` to Screenshot but also `` Hyper + Left Shift + ` `` to a separate keybind if I wish as Karabiner Elements distingushes between left and right modifiers (screenshots below show it as record screen) 

Can I do similar in Kitty? I'd like to open up modifiers for the hyper key and add `map` keybindings that support knowing that specifically the right modifiers were pressed and not just the modifier itself like Karabiner. 

Is something like `map right_control + right_command + right_shift + right_option + left_shift + w swap_with_window` supported in kitty.conf?

![image](https://user-images.githubusercontent.com/13072760/181631062-2db74d4b-2979-45e9-85ae-4ff2e0c10ac6.png)
![image](https://user-images.githubusercontent.com/13072760/181631195-94423f05-395f-4e09-87dd-34bbbe144723.png)

**Comment by alaricljs** [comment, marked as answer]
It's in [the docs](https://sw.kovidgoyal.net/kitty/conf/#keyboard-shortcuts) `kitty --debug-input` will tell you what you need for key codes.  Unfortunately on Mac, I think you're stuck with glfw hex codes in order to differentiate between mods.   I haven't validated that this works on the mod side, but it's easy for you to try it.

**Reply by roscrl** [reply]
Ah thank you it looks for right_control I get.. 
![image](https://user-images.githubusercontent.com/13072760/181638607-bc9f10ee-218b-4dd2-9980-35820a8ac4db.png)

And for left_control..
![image](https://user-images.githubusercontent.com/13072760/181638741-f070a8b7-822a-48d8-bab5-a069d394c03f.png)

I assume I can somehow use the 0xe068 (right) vs 0xe062 (left) distinction (hopefully same for other modifiers), will continue to read docs.

**Reply by alaricljs** [reply]
`map ctrl+0x61 something` is the example in the docs.  If it's going to work as a modifier then it's just:
`map 0xe068+w send_text blah` and you try right control and w and see if you get 'blah' spit out to the terminal.

**Reply by roscrl** [reply]
Hmm no luck with that. I tried 
![image](https://user-images.githubusercontent.com/13072760/181642333-2920068c-5c77-419d-bc16-af1f1812207b.png)

Using cmd+j as a point of comparison.
~~I am just noticing in the input that it says `ignoring as keyboard mode does not support encoding this event` wonder if that could be an issue.~~ That happens on normal left control also.. ignore me.

**Reply by roscrl** [reply]
Also no luck with trying the native key like `map 0x3e+j set_tab_title`

**Reply by alaricljs** [reply]
Still have not found the code that is involved.  I can say that it also doesn't work in Linux/X11, so even tho X11 explicitly reports named keycodes for both XKB and GLFW.  They don't work as modifier names, only the explicitly documented names work.

**Comment by kovidgoyal** [comment]
kitty itself can, as can be easily seen with

kitty kitty +kitten show_key -m kitty

and pressing left and right modifiers.

However, IIRC, mapping does not distinguish between left and right
modifier.

**Reply by page-down** [reply]
I think this should be the correct answer to this question post.

This feature request was discussed before.
https://github.com/kovidgoyal/kitty/issues/4908#issuecomment-1085721014

As you learned, kitty recognizes left and right modifier keys, however it takes time and effort to get the key mapping to recognize them and ensure that it works properly on both macOS and Linux.
